### DISCUSSION TITLE
Helix/SP Builder user wants to build with “Astroid”

### DISCUSSION THREAD

**Discussion by paurray** [templaza/astroid-framework#156] [Q&A] [answered]
Hello (First time poster here :-)

Having watched several hours of tutorials about Astroid I have a decent grasp of how the Framework works.
Seems similar but better to “Helix” with out the “SP Page Builder” both of which I have used for 5 years +/-.
With SP Page Builder I would create SP Pages (place things on the pages) and link up to Joomla Menu items.

What my brain thinks should be happening in “Astroid” is that I….

a) Create pages & tweak pages.
b) Link Pages to Joomla Menu items

As far as I understand things there was a  "JD builder" that has now also been discontined.

(I just don’t get it… I have a decent Joomla experience 10 +/- years but not good with code/overwrites etc)
Is it possible to use “Astroid” **to build sites** with out learning to code & if so could someone kindly point me in the right direction?

thanks

Paul

**Comment by Chacapamac** [comment]
I @paurray 
**I will try to help.**
I’m not affiliate with the NEW Astroid and long time Joomla Websites develloper. I discovered Astroid (when develop by Joomdev) a few years ago when I wanted to change  from my old, trusted Rockettheme and their new Gantry framework that didn’t fit anymore what I see as a Good Joomla Friendly Framework.

 For more than a year, I tested all the bigs Joomla Framework out there (including Helix) & Astroid win my vote by far.  I’m happy to give some time _(when I can)_ to the even better framework the NEW Astroid is become.

**You said:**
> a) Create pages & tweak pages.
> b) Link Pages to Joomla Menu items

Well, this is exactly how Joomla work. Astroid use the full native power of Joomla. See Astroid as a helper to build and modify easily the “Framework” and design of your Website with it’s great template manager + all the extras; articles Types, SEO, Control of typography, design, responsive design, types of Headers and all the other great stuff...

**You said:**
> As far as I understand things there was a "JD builder" that has now also been discontinued.
Yes it was and I think, I need somebody to confirm, that **Joomlaplates** ( https://www.joomlaplates.com/) will or have taken over the continuity of the existing Joomdev Templates and JD Builder and other JoomDev extensions.

In my year testing all the top Joomla frameworks, I also test all “Builders” that each company had and I wasn’t impressed (at that Time) — You can see my impression of them at that time on Joomla forum —> https://forum.joomla.org/viewtopic.php?f=575&t=988600

**In Fact, To create the content with the tandem Joomla/Astroid, you can use the Editor/Builder you want and fit your need.**

—————————————————————————————————
**For me, my only 2 necessary criteria for a great Joomla Page Builder Editor was:**

**1. TOTAL INTEGRATION with the native Joomla system.** 
(articles manager, articles edition & custom modules edition, media's manager, categories...) and should present and HYPER IMPORTANT should have an IDENTICAL Joomla edition experience both from frontend & backend.
— This is important! Anything else will complicate the overall CMS experience to the detriment of Users/Editor and diminish the power of Joomla itself
— One Native Joomla content management & Edition system to learn
— Should work just like JCE or any other editors in complete tandem with Joomla FIRST (They can add gizmos if they want)
— Remember, we have to beat the competition (WP Gutemberg)

**2. Being simple & intuitive to use and at the same time as powerful as a traditional editor like JCE.**
— Also IDEALLY made by a Joomla multi developers renown establish business.

Any Page Builder that doesn’t comply with these 2 criteria is simply NOT “Joomla Friendly” and not using the full power of Joomla properly and detrimental to Joomla itself.
—————————————————————————————————

Personally, for now, I will continue to use JCE as my editor of choice, with its bootstrap columns ability, the custom template and styling ability, the clean HTML code generated and the total control of the code (with its powerful collapsing tags and searchable code editor), the ability to show a WYSIWYG view of your edition, the incredible set of tools that you can have or decide to limit depending of the user. I’m sticking with it for now.

**I also know that **SP Page Builder** is now more flexible and will probably be my choice.**
— I like the frontend edition
— I like the fact that you can use it with native Joomla articles (backend/frontend) and contrary to what **Joomshaper** says this is not a limitation because with Astroid, you can easily build a layout where the Joomla Component area (where Joomla articles appear) is Full Width when you need to build a front page of for any other page where a full with visual is needed. Joomshaper do that to force you to use their Component “Pages” interface bypassing the powerful Joomla native system that you will need anyway to integrate with other systems and forcing you to learn two different interfaces (Native Joomla and SP PageBuilder Pages Component) to create the same content.

Using my little trick, You can use SP PageBuilder in complete integration with the native Joomla system and, if I’m right, they even have a special custom module that work with SP. That way, you get a drag and drop Editor that work perfectly with Joomla. One way to edit everything.

**Comment by paurray** [comment]
@Chacapamac Thanks for your detailed response.

I read this article recently:
https://www.dionysopoulos.me/joomla-performance-tuning-i-start-at-the-beginning.html

And decided to go back to the start after building crazy complex sites that got out of control.

After a decade of Joomla I have a lot of really bad habits as in reaching for big fat components before getting the basics down. Thanks for reminding me of the JCE Editor. I will install it and look at it again. Always seemed very down to earth to me. I have used SP Pro with Helix (many other components good/bad & ugly as well) for years.

Specifically with respect to SP Pro I have a love/hate thing with it. Seems to me to be...

- Clunky (never used the latest version/ultimate or what ever it is called)
- Buggy (things breaking with updates)
- Support was patchy (If I am going to use it I would pay for the Pro version & patchy support ruins it for me :-(

+ Agree front end editing is very cool particularly if you want clients to play ball.

Without wanting to get into a pros & cons template discussion ;-)
Astroid really impresses me on many levels.

+ Mega Menu
+ Article options seem amazing.
+ Adding videos & pimping video thumbnails
+ Complete control over just about every aspect of the template (no code skills required :-)*
* I am building a very simple site for personal use (hero page etc) but hope that if I get me head around Astroid that I can use it for pretty much anything else in the future. This is for me the single most attractive aspect of Astroid :-)

- I do find accessing “Astroid” a little weird as in going to Templates. But yes it makes sense.
(Have now put a “Template Styles” button short cut on the “Joomla Home Dashboard” in the “System” section)
- Loading screen with “astroid” drives me nuts. But that is just me!

But back to getting Astroid working with Joomla (tutorials I am watching are all out of date)….

Understanding the workflow so far...

1) There are pre loaded modules that you can configure and add to all web pages or switch off for some.

a) Header
b) Social Icons
c) Branding
d) Contact Info.

2) That these pre designed modules live in “Sections” of the Astroid Layout manager. Can be tweaked & turned on & off.

3) That you can add an infinite number of sections and populate them with for eg Joomla custom modules (little + button in the middle of the Astroid sections on the Astroid layout page).

4) Currently trying to get my head around module positions for custom modules combining Joomla & Astroid.
I can get a custom module to go to the banner section which is a one off pre loaded Astroid thing. Bbbbbbut once I start inserting my own Joomla use other module positions like left/right or content top I get them showing up twice.

Need to get this knot out of my brain before moving onto different modules on different pages…

![Site-MockUP-22](https://user-images.githubusercontent.com/9163424/182120142-197c0fe2-373c-447e-9a64-3cc6bc2c6132.jpg)
.

**Comment by paurray** [comment]
Update this video here:
How to add a module position
https://www.youtube.com/watch?v=1MxREZ0u3N0
Gave me the clue I needed. To get step 4 right.

I need to create custom module positions by simply adding them in the Position filed of the custom module. I probably knew at some point. But forgot about it :-(

4a) I created a custom hero position 1 & 2 in custom modules in Joomla
4b) And assigned & select this same custom positon in an Astroid section/custom module.
Hope that makes sense.

Next up Step 5 getting different modules for different pages!

![MockUp-Home02](https://user-images.githubusercontent.com/9163424/182168829-d609d63c-c8a3-4002-9e57-1bb3b0948360.jpg)

**Reply by Chacapamac** [reply]
Yes, this one of the great force of Joomla, it’s flexibility...
**Simple as that**

**Yes JCE is hard to beat**
Some of the features I Love!

**• One of the best new stuff in JCE is the ability to create responsive collapsing columns**
https://www.joomlacontenteditor.net/support/documentation/columns/creating-and-editing-columns

![Screen Shot 2022-08-01 at 11 22 48 AM](https://user-images.githubusercontent.com/1907276/182185854-e59a1c33-d54d-480d-9ade-ac0be0a0c26c.jpg)


**• Create easily applicable complex custom Styles**
https://www.joomlacontenteditor.net/support/tutorials/editor/creating-custom-styles


**• JCE can  be a completely WYSIWYG editor with its basic by default use of all the css files in your template > css folder with powerful custom settings.**

_**Note:**_ If you want to add the ability to see Awesome icons in you editor you can add something like —> @import "../../../media/astroid/assets/vendor/fontawesome/css/all.min.css";
In your Custom CSS Files
![Screen Shot 2022-08-01 at 11 57 12 AM](https://user-images.githubusercontent.com/1907276/182190076-cf3bdc6e-6837-4588-b07d-bdade3815939.jpg)

**• and also the powerful ability to insert complex pre-prepared simple or complex templates in one click**

and so much more......




**Do You have other questions on the custom modules creation** @paurray **?**

**Comment by paurray** [comment]
Hi @Chacapamac 

Joomla is great sometimes however the sheer number of possibilities can be overwhelming. This is one of the reason I am trying to keep things as simple as possible for as long as possible. My guess is that this is the main reason why people give up on Joomla and move on to lesser Content Management Systems ;-)

Thanks for more details about JCE.
I will look at in more detail when I am done with the basics.

Yessss thank you I could really do with some help with respect to nailing the menus using standard Joomla stuff. My thing number 5 to get done. The whole menu thing. Again I have another knot in my brain…

I set up a main menu (english).
The home page is set to Featured Articles which I guess in retrospect is a mistake. Maybe should have built menus right first :-(

I watched this video here which is in German. (I understand German.)
“Megamenu Konfiguration ”
https://www.joomlaplates.de/template-konfiguration/megamenu-konfiguration.html
Where the Gentleman doing the video mentions (@28 sec) that the main menu needs to be a “trennzeichen” which I think has to be a delimiter/separator in Joomla english. Assuming that this is correct???

I am in a stalemate situation at the moment.

a) Going to my existing “Home” menu.
- Menu Item Type & clicking “Select”
- Going to System Links”
- Selecting Separator
- Clicking Save

And I get the following error:

“Save failed with the following error: The home menu item must be a component.”

b) Creating a new menu “Home tk2 EN”
- Menu Item Type & clicking “Select”
- Going to System Links”
- Selecting Separator
- Clicking Save

Is all good.

Butttt

- Selecting “Home tk2 EN” in Main Menu (EN) (The main menu for the site)
- Going to the Actions drop down menu
- Selecting “Make Home”

And I get the following errors:
“The home menu item must be a component.”
"No menu item set to home."

So I am stuck.
I have done a couple of hours of googling but am lost.
Can you kindly give me a nudge in the right direction?

Any help would be greatly appreciated.

01 multilingual site with EN as the default language

<img width="720" alt="001-finalBUGTV-Menus-Positions" src="https://user-images.githubusercontent.com/9163424/182333160-f5b16fad-b423-4da5-8d6a-f9a838e2b293.png">

02 How things are set up now with 2 attempts to nail home page

<img width="720" alt="002-finalBUGTV-Menus" src="https://user-images.githubusercontent.com/9163424/182333479-65402ba7-120a-43c6-9907-d95cf77fd0f9.png">

a) The exisitng menu attempt to change to "Separator"

<img width="720" alt="003-finalBUGTV-featured-Art" src="https://user-images.githubusercontent.com/9163424/182333820-abc9de57-b278-4af7-adfc-1b4e22cb6584.png">

b) having created a 2nd "separator" menu item and attemting to make it a home page via Actions menu

<img width="720" alt="004-finalBUGTV-Separator" src="https://user-images.githubusercontent.com/9163424/182334160-589562ff-741d-435b-af25-4e394c667ac2.png">

**Reply by Chacapamac** [reply]
I think you cannot put the home menu type as a delimiter/separator. Just like the error say it need to point to something; a component, a single article or Featured article (even if you have no Featured article set)

**Why do you want a Separator as you menu item?**

Separator are normally use as a no content main menu item to choose/categorize sub-menus.

**Comment by paurray** [comment]
Hi @Chacapamac

Yes my bad. Sorry the Separator idea was rubbish. I was getting mixed up between formatting a Mega Module and getting a basic menu up and working. The last thing I need for my simple site is a Mega Menu. Please ignore the last section from me above.

As also mentioned previously I have been contaminated by way to many fancy components. A case in point being “Easy Blog” from Stackideas which I love don’t get me wrong. But my bad fluffy habits die hard. I have never actually used Joomla´s Blog capabilities. Ever!

Any way I switched from Joomla Blog Category to Single Joomla blog Article in the menu type options for the home page and other pages as well. And posted a blog called “Dumb Article” with the content “Dumb Article” which I can hide later. I now have my 2nd page up and running with different modules going to it formatted using the Astroid Template Frame work. So I am definitely on the right track.

The solution does seem strange to me with the single article option and then hiding the article but hey what would I know. I will play with this for a couple of days and see how it goes. I  will write a summary of what I did in case it helps any one else…. Maybe you want to look over it and correct any mistakes I might make. Astroid rocks. Sorry about the pun but it is true. If it remains freely available and is developed further that would be awesome.

<img width="720" alt="finalBUGT-Menu-Settings" src="https://user-images.githubusercontent.com/9163424/182470659-e62395ce-2372-4202-a3ab-500e473d5833.png">


<img width="720" alt="finalBUGTV-2nd-page" src="https://user-images.githubusercontent.com/9163424/182470017-2c744c42-2cdd-45e6-a8a7-dc5b4e0457a4.png">

**Reply by Chacapamac** [reply]
I understand You!
> As also mentioned previously I have been contaminated by way to many fancy components. A case in point being “Easy Blog” from Stackideas which I love don’t get me wrong. But my bad fluffy habits die hard. I have never actually used Joomla´s Blog capabilities. Ever!

Personally, I try to do the most I can with Joomla first before bringing a component in the mix. I always work with the native Joomla blog option and maybe adding some plugins for added fonctions, if needed.

**When you said;**
> Any way I switched from Joomla Blog Category to Single Joomla blog Article

In Joomla you don’t have a **_Single Joomla blog Article_** menu item, you have a **_Single Article_** item.
— If you want a blog, you take the “**Category Blog**” menu item and if you want a single page/article you take the “**Single Article**”

The **Category blog** item give you something like that —> https://efoilsurf.ca/en/blog-events-news-efoilsurf-canada
— You just create the articles in the categories assign to the **Category blog** item, that’s it...

**Comment by paurray** [comment]
For my future self and any one else who might be interested, disclaimer, I do not really know what I am doing…

- ***Probably makes sense to work in a copy of a template and back up first etc***
- Also found this helpful: https://docs.joomla.org/Module_Position


**Get some Joomla pages/menus up and running that are essential blank/place holders.**

1) Build standard Joomla Menus (eg going to dumb articles/categories). Something that you can hide. You just want a place holder.


**Creating Content for Specific Menus/Pages in Joomla**

2a) Add Content by creating Site Modules (eg Custom Modules) using standard Joomla Site module manager.

2b) Assign Modules to Joomla pages using standard Joomla module manager. It is important to realise that you can give Modules a unique position names like home left & home right or about header about text etc. At this point you are only preparing the modules for exact placement later. (Steps 2c & 3b & 3c) Give the positions names that you will understand later. Do so in the position field on the right hand side of the first page in the module editor.

- Don’t be bothered by the fact that Joomla does not recognise your module name. Just type and when you are done hit enter/tab and then Save.

- Also you need to be very careful at this point as Joomla tries to be cleverer than you and correct you like an auto spell check and change your unique position name to something generic like left!

2c) Assign the module to the page you want it to appear on in the Menu Assignment page. Again it will only actually get assigned afterSstep 3b & 3c later.

2d) Make sure the module is published.

2e) Having gone to the trouble of nailing the above you can always use the Save as Copy function in the drop down menu under the point Save and Close (top left) to get things done more efficiently and potentially making fewer mistakes. As in give copy a new name and change where appropriate.

EG currently building 3rd page (Where EN is english)

With the following Joomla Modules:

03a EN Work (Image Films-Adverts)
03 EN Work (Television-Web)
03 EN Work (Doc-Social-Media)
03d EN Work (Drama-Music)

Each with the following as defined by me position:

03a EN Work TV-Web
03b EN Work Image-Ads
03c EN Work Doc-Social
03d EN Work Drama-Music

All on the following (selected only) page:

03 Edit Work (third page on my site)

Here are some screen shots for getting things working at the Joomla end...

<img width="720" alt="A-Another-Joomla Detail" src="https://user-images.githubusercontent.com/9163424/182838856-9a48c1af-2a26-499d-ade1-dba3258d53b4.png">

(Above) Inside a Custom Joomla Module. Adding content. Setting custom Module Position & naming stuff.

<img width="720" alt="A-Also-Module-Position" src="https://user-images.githubusercontent.com/9163424/182828804-6807cb63-f849-4a73-a0de-957280658004.png">

(Above) Have made sure that each custom Module with custom module position is set too desired page ONLY.

<img width="720" alt="A-Setting-Up-Modules-Joomla" src="https://user-images.githubusercontent.com/9163424/182828872-7c7361e3-8353-44f4-b3a8-f4a4fa0ee813.png">

(Above) Have assigned Custom Modules to custom module positions. Nothing showing anywhere yet.

**Putting the content in a specific place on a Menu/Page with Astroid Layout Manager**

3a) Obviously you have the Astroid Frame Work & at least one Template installed. Navigate to the Astroid Layout setting of your Astroid Template.

3b) In the Astroid Layout section you need to first create and configure a section.

EG Continuing with my 3rd page:

Select “new Section” which you find above one of the existing sections. (Technically you could put the section anywhere on the Astroids layout page. I am placing it between the Section after (01) “Home” & (02) “About” and before the Section (04) “Work” which I will make later….)

But back to the current Section which will be displayed on Joomla Menu/Page 3 when populated with Joomla modules pointing to Menu/Page 3!!! Having selected New section then Select module and select a Grid layout. In my case I am selecting 3 3 3 3. For my 4 modules. Give the section a name that makes sense.

EG Astroid Section Edit Work Page

3c) Click the Plus button in the first row of the Astroid Section Edit Work Page and type the name of your custom module position. Name, make design tweaks in the design tab and Save the change to the Element containing a Joomla module in the Astroid Section and save bottom right of dialogue box. Save also the change to the layout page in astroid at top left of Astroid Layout page.

Note. Select Edit Element (top right of Module in Section) if you need to go back and tweak again. Also be careful you might have cache issues switching from Joomla to Astroid. You might want to Clear Cache & Reload pages etc. 

Here are some screen shots for getting things working at the Astroid end...

<img width="720" alt="C-Lyout-In-Context-Astrroid" src="https://user-images.githubusercontent.com/9163424/182829931-1915f304-3d62-4647-a754-b5bcaa3f0e03.png">

(Above) Have set up a section on the Layout Page in Astroid will use this section to load the custom Joomla Modules.

<img width="720" alt="D-Assigning-Joomla_Module-HEre" src="https://user-images.githubusercontent.com/9163424/182829981-f3951b0b-12ab-47d5-a641-62fa35cf99ef.png">

(Above) Inside a Section inside a row inside and added Astroid Module that point to the Custom Joomla Module/Position!

Here is a screen shot of how things are starting to look on the Joomla front end...

<img width="720" alt="E-Result-Edit-Work-Page-rk2" src="https://user-images.githubusercontent.com/9163424/182880456-b537c174-4af1-4390-93f3-903d59c9dd4e.png">

(Above) The result with my Dumb Article that I used to set up page. Maybe not my brightest idea. Who knows.

Next steps...

Having got all the knots out of my brain. And actually building something I am finding that the trickiest thing is some time having to clear the cache to get Joomla & Astroid to work hand in hand. Often just refreshing the Astroid page seems to do the trick. In all fairness Cache is always a bit tricky in my experience.

Next up I will

- Finish my English Site Structure
- I want to nail as many of the formatting options as possible in Astroid as in make a template.
01 COLOR THEME (Might involve sass which scares me)
02 TYPOGRAPHY
03 BUTTONS (No idea really where to start)

- Look at JCE and see how much sense it makes. The Tiny WYSIWYG editor is already beginning to get on my nerves. I can imagine that JCE it makes formatting even more granular. Maybe doing more boxes inside boxes inside Joomla Modules rather than more Joomla modules inside Astroid layout sections. If that makes any sense. But that is a task for next week…

@Chacapamac  if there any glaring mistakes in above please feel free to chip in. Thank you. I really do not pretend to know what I am doing. I will probably spot mistakes and update minor things above at the latest next week.

**Comment by Chacapamac** [comment]
I @paurray, I can see that you are working hard on that.....

**You said:**
> Probably makes sense to work in a copy of a template and back up first etc

**This is the basic rule of professionally making a website with any CMS.**
_• Before doing any changes, before installing anything backup your work first._ 
Using Joomla or Wordpress, I use the ABSOLUTELY necessary component free version of **Akeeba Backup** (also available for WP)
**Download** —> https://www.akeeba.com/download.html
— You need also the free **Akeeba Kickstart** (on the same page) to restore your backups if necessary.
_**Note:** The restore process is even easier with the commercial version of Akeeba Backup._

**My personal strategy, when a website is online:**
I keep a local clone with both website equip with Akeeba Backup. I NEVER do an update of Joomla , extensions or code modification on the live website without testing everything on the local clone first and port the change on the live site after verification. All that with backups on boths websites before modifications and after modifications.
This technique save my skin at multiple occasions.

**Comment by paurray** [comment, marked as answer]
Update. Just wanted to add that I am now using JCE Editor Pro (thanks for the tip @Chacapamac ) and love it in combination with Astroid. All very intuitive and straight forward having learnt how it all works together!!! Amazing tools, resources & community. Thanks.

- Doing misc colours, fonts etc in Astroid and…
- Building pages in Joomla modules
- JCE for columns & bare bones page layout in Joomla Modules.
- Putting Joomla Modules in an Astroid Layout Section that I can use to tweak formatting eg backgrounds etc.


<img width="960" alt="finalBUGrv-FINISHED" src="https://user-images.githubusercontent.com/9163424/189499773-da006351-04a1-4d20-b8a8-62fca631a6d9.png">

**Comment by Chacapamac** [comment]
JCE is building absolutely clean minimalist html and is a WYSIWYG editor with so much power. It can be make simple for authors/editors. For me it beat all those Drag & Drop builders as you can do the same with the Templates and Styles. Worse, these Builders create complex, buggy html and their content break down as you update them. Also, so far, none are using the full power of Joomla where you can edit, with the same editor in backend and frontend and get the full flexibility of the article manager and also edit custom modules front & back 
— NEWS — Ryan of JCE is also working on the ability to drag and drop element (div, p section...) in the editor window to facilitate positioning.

Oh, I maybe forget to say that with the flexibility of Astroid you can easily create a full front page with only one article (main Joomla component) or one custom module by using a special layout for you front page.
